### ISSUE TITLE
Resolve issues with Indiana monitor

### ISSUE THREAD

**Issue by alanjosephwilliams** [codeforamerica/snap-it-up#49]
The "FSSA" in "FSSA Benefits Portal" has been removed from the page, so we are showing continuous
downtime erroneously.

**Comment by alanjosephwilliams** [comment]
We've updated the string to  "Apply for Health Benefits, SNAP"—the thinking being that this active link would only be displayed if the service was unavailable. However, @pingometer is still reporting Indiana is as down. Investigating now, along with the Vermont issue.

**Comment by pingometer** [comment]
@alanjosephwilliams It looks like their site is blocking our checks. We only receive this as a response: `<html><head></head><body></body></html>`. To be safe, we tried "faking" a few different user agent strings and it still doesn't work, so they must have another mechanism in place which is stopping it out.

**Comment by Mr0grog** [comment]
Oof, the way that site works is nuts:
1. HTTP redirect to `/CitzenPortal/userhome.jsp`
2. _JavaScript_ redirect to `/CitizenPortal/application.do`
3. Load pretty much all the page content via XHR and JavaScript.

@pingometer I can move through that whole process with curl (save the JS redirect), but I never get a totally empty page. It still has JS and a tiny bit of `<noscript>` content in it. It seems like you have to keep cookies to get all the right redirects, though.

**Comment by Mr0grog** [comment]
@pingometer for transactional checks, do you automate an actual browser (e.g. via Selenium) or are you emulating w/ something like, say, mechanize?

**Comment by Mr0grog** [comment]
@alanjosephwilliams There’s maybe something valuable in that for the narrative: site that requires JS to work at all -> contributes to the slowness, also may not work or at least work well on low-end devices.

**Comment by pingometer** [comment]
@Mr0grog It looks like `/CitzenPortal/userhome.jsp` throws a 404 error when accessed directly, so we can't test against that. We use Selenium, with full support for client-side script, cookies, and an implicit wait of 10 seconds for XHR/AJAX. Are you able to screenshot it with Selenium?

**Comment by Mr0grog** [comment]
@pingometer I recently took down our own server that we were using for automation and screenshots, but using Browserstack (which, as far as I understand it, is selenium under the hood), it seems to work. Don’t know what special sauce and configuration they might add in, though.
http://www.browserstack.com/screenshots/e10abd9ea5cb7382954b530c872aeb5db419887a

**Comment by pingometer** [comment]
It's possible that it's related to using Firefox (instead of Chrome or PhantomJS which we use). We'll run some tests.

**Comment by alanjosephwilliams** [comment]
> site that requires JS to work at all -> contributes to the slowness, also may not work or at least work well on low-end devices.

@Mr0grog definitely going to note that in our ongoing list of "basics"

**Comment by Mr0grog** [comment]
@pingometer are you using Chrome or Phantom? One thing we noticed is that several sites can’t be loaded in Phantom because they have SSL errors (!) and a bug in Phantom prevents you from overriding those errors. See https://github.com/codeforamerica/snap-it-up/issues/11#issuecomment-77018195 and https://github.com/ariya/phantomjs/issues/12181

Most shipping consumer browsers (e.g. Chrome works but not Chromium/Phantom) are more lenient and work even though there is an SSL error, so automating them works. SlimerJS (similar API to PhantomJS, but fore Firefox) doesn’t have the you-can’t-ignore-ssl-errors bug Phantom does, so it also works.

**Comment by pingometer** [comment]
@Mr0grog Yes, we use PhantomJS 99% of the time and encountered the SSL bug in a few cases. It also doesn't have support for disabling alert boxes that are fired when the page loads, so we have some crafty workarounds to mute them. We're hoping (praying) for that bug to get fixed.

**Comment by alanjosephwilliams** [comment]
@pingometer @Mr0grog just to confirm—we do not as of yet have a remedy for the Indiana monitor? Any results from your tests @pingometer?

**Comment by pingometer** [comment]
@alanjosephwilliams @Mr0grog We were able to build a custom version of the browser to tolerate the issues (namely redirects/failed SSL handshakes) with this site. It appears to have resolved the issue (checks are coming in as up), although the site does timeout on occasion (we wait for 45 seconds and it's hitting that).

**Comment by Mr0grog** [comment]
@pingometer sweet, thanks so much!

**Comment by alanjosephwilliams** [comment]
@pingometer we are still reporting Indiana as down. It sounds like timeouts are the likely reason from your end, but we are currently seeing no response time at all from the monitor.
![image](https://cloud.githubusercontent.com/assets/2983464/7030231/f5edf758-dd17-11e4-8c51-751d46117164.png)

I've updated the title of this issue to reflect the current concern. Ideally, how we address the problems with this monitor can inform a consistent QA strategy going forward.

**Comment by pingometer** [comment]
@alanjosephwilliams If a check times out, the response time is reported as 0 since we don't really know how long it would've taken for the server to respond. It seems like a 50-60 second timeout allows checks to proceed for this monitor, but that's extremely long for a site to load. Thoughts?

**Comment by Mr0grog** [comment]
**TL;DR Personally, I think the way Pingometer is behaving right now is ok, but I realize there is room for a range of opinions there. We need more people than just me to weigh in on what’s reasonable (that means putting your vote on an actual number).**

So… **technically** there’s no inherent time limit in HTTP, but most browsers [w/ default settings] give up around the 2 minute mark (couldn’t find an exact number for Firefox, though I’m sure it’s out there; experimentally, Safari seems to be hanging on a bit longer, but I can’t get a number out of it; Chrome is right at 2min).

**Practically speaking,** though, > 10s is generally considered slow and > 20s is pretty darn unreasonable these days. It’s questionable whether one should expect a commercial service like @pingometer to wait two minutes. On the other hand, we could be called disingenuous if we don’t wait as long as a normal user’s web browser would (we’re kind of a special case, as we’re monitoring _other people’s_ services _and_ we’re doing it to hold them accountable—very different for what a service like Pingometer is normally designed for, which is let yourself know when things aren’t going right on _your_ service). On the other other hand, it would be pretty ridiculous to expect a user to wait that long for a site to load. Just as we could be called disingenuous for not waiting as long as possible, we might also be more realistic or practical to cut it at something like 30-40s. I personally fall more towards that argument.

It would be awesome if we could configure the timeout value for our monitors (say, within the range 10-120s). It would also be pretty neat if Pingometer could represent “intermittent” connectivity events instead of just up/down. All of these things might also be totally unreasonable for them to do, though. Again, we represent a lot of weird, fringe needs and a very uncommon use case. What they’ve already done to accommodate us is pretty darn fantastic. (Thanks, @pingometer folks!)

**Comment by alanjosephwilliams** [comment]
Seriously the support from @pingometer has been tremendous.
- Given the current approach, its almost impossible for us to present credible data about the uptime of a monitor like Indiana's.
- Something like an arbitrary-but-reasonable-sounding cut off like "one full minute" seems to me to be more defensible than 40 seconds—at least to a lay audience, if only because it is easier to conceptualize. I think browser parity is also defensible and easy to conceptualize, if overly generous.
- The ability to configure the timeout would be a tremendous feature for our purposes—lifting a significant QA onus.
- Nonetheless, our most pressing need is a manual QA process—something like manual verification of a monitor that has been down for more than a given period of time.

**Comment by Mr0grog** [comment]
> Given the current approach, its almost impossible for us to present credible data about the uptime of a monitor like Indiana's.

Side note: there’s an interesting anecdote in this fact alone. The site is mind-bogglingly slow, it has SSL verification problems (!!!), pages literally _do not load without cookies,_ and it doesn’t work (at all, you just get a blank screen) without JavaScript. It’s kind of a nightmare—and in plenty of ways that have nothing to do with monitoring.
